### DISCUSSION TITLE
Strip hash comments from a TypeDefinitionRegistry

### DISCUSSION THREAD

**Discussion by seanf** [graphql-java/graphql-java#2765] [Q&A] [answered]
Is there some way I can strip all hash comments from a schema in the form of a `TypeDefinitionRegistry` (or a `graphql.language.Document`)? I want to preserve descriptions, but not hash comments.

**Comment by seanf** [comment]
This is what I did in the end:

```java
    AstTransformer astTransformer = new AstTransformer();
    CommentStripper visitor = new CommentStripper();
    Document documentWithoutComments = (Document) astTransformer.transform(documentWithComments, visitor);
```

The class `NodeCommentStripper` extends `NodeVisitorStub` and the method `visitNode` calls `transform` on each `Node` to replace the comments with an empty `List`. To avoid implementing 36 virtually identical visitor methods (and hopefully handle any future `Node` types), I used reflection when calling `transform` and `Builder.comments`.

Ref: https://github.com/graphql-java/graphql-java/pull/1335#issuecomment-443476271

**Reply by seanf** [reply]
In the _actual_ end (so far), I did this: https://github.com/graphql-java/graphql-java/discussions/2765#discussioncomment-2411123

**Comment by bbakerman** [comment, marked as answer]
We did some work a while back to separate `#` comments from descriptions.  Hence the `Description` class versus the `Comment` class

However we never put in a parser option to say "dont collect" them.  In theory you can write an AST transformer that could get rid of them but I think a parser option to never collec them in the first place is better

https://github.com/graphql-java/graphql-java/issues/2767

If you are using SchemaGenerator for building your schema can set `graphql.schema.idl.SchemaGenerator.Options#isUseCommentsAsDescription` (which defaults to true) so setting that to false would mean that comments would not come into the "element descriptions"  This flag as left as true during the migration from #comments to proper descriptions.


Can I ask - why is it that you dont want them?  Is it schema printing?  memory used?

**Reply by seanf** [reply]
> In theory you can write an AST transformer that could get rid of them but I think a parser option to never collec them in the first place is better
> https://github.com/graphql-java/graphql-java/issues/2767

Agreed! Thanks, I'll keep an eye on that issue.

As for why, some background is [here](https://github.com/Netflix/dgs-framework/discussions/939#discussioncomment-2410946), but yes, I need to print the schema (after it is augmented by DGS), and I found that the hash comments were being printed as descriptions. And I don't really want the hash comments anyway, because they're not really intended for the users of the schema.

**Reply by seanf** [reply]
Thanks @bbakerman, I was just missing the option to `SchemaGenerator`. I should have paid it more attention, but I hadn't noticed that it was on by default. It seems like a weird default, but maybe this behaviour was needed before String descriptions were in the spec? (I guess that's what you meant about the migration.)
  
Since I only need to remove hash comments from the final SDL, this was all I needed in the end:

```java
    SchemaGenerator schemaGenerator = new SchemaGenerator();
    SchemaGenerator.Options schemaGenOptions = 
        SchemaGenerator.Options.defaultOptions().useCommentsAsDescriptions(false);
    GraphQLSchema graphQLSchema = schemaGenerator.makeExecutableSchema(
        schemaGenOptions, myTypeDefRegistry, RuntimeWiring.newRuntimeWiring().build());
    String newSchemaSDL = new SchemaPrinter(SchemaPrinter.Options.defaultOptions()).print(graphQLSchema);
```
